File: Data Structures & Algorithms/merge-intervals/submission-6.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:

        if len(intervals) == 1:
            return intervals

        result = list()
        intervals.sort()
        
        for i in range(len(intervals) - 1):
            if not result:
                result.append(intervals[i])
            if result[-1][1] >= intervals[i + 1][0]:
                result[-1] = [min(result[-1][0], intervals[i + 1][0]), max(result[-1][1], intervals[i + 1][1])]
            else:
                result.append(intervals[i + 1])
        
        return result

File: Data Structures & Algorithms/insert-new-interval/submission-6.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> result = new ArrayList<>();

        for(int i = 0; i < intervals.length; i++) {
            int[] interval = intervals[i];

            if(interval[1] < newInterval[0]) {
                result.add(interval);
            }
            else if(newInterval[1] < interval[0]) {
                result.add(newInterval);
                while(i < intervals.length) {
                    result.add(intervals[i]);
                    i++;
                }
                return result.toArray(new int[result.size()][]);
            }
            else {
                newInterval[0] = Math.min(newInterval[0], interval[0]);
                newInterval[1] = Math.max(newInterval[1], interval[1]);
            }
        }

        result.add(newInterval);
        return result.toArray(new int[result.size()][]);
    }
}
